File: packages/sdkwork-utils-java/src/main/java/com/sdkwork/utils/ResultUtils.java
package com.sdkwork.utils;

import java.util.function.Function;

public final class ResultUtils {
    private ResultUtils() {
    }

    public record ResultValue<T>(T value, String error, boolean ok) {
        public static <T> ResultValue<T> ok(T value) {
            return new ResultValue<>(value, null, true);
        }

        public static <T> ResultValue<T> err(String message) {
            return new ResultValue<>(null, message, false);
        }
    }

    public static <T> boolean isOk(ResultValue<T> result) {
        return result.ok();
    }

    public static <T> boolean isErr(ResultValue<T> result) {
        return !result.ok();
    }

    public static <T> T unwrapOr(ResultValue<T> result, T defaultValue) {
        return result.ok() ? result.value() : defaultValue;
    }

    public static <T, U> ResultValue<U> map(ResultValue<T> result, Function<T, U> mapper) {
        return result.ok() ? ResultValue.ok(mapper.apply(result.value())) : ResultValue.err(result.error());
    }
}

File: packages/sdkwork-utils-kotlin/src/main/kotlin/com/sdkwork/utils/ResultUtils.kt
package com.sdkwork.utils

data class ResultValue<T>(val value: T? = null, val error: String? = null, val ok: Boolean = false) {
    companion object {
        fun <T> ok(value: T): ResultValue<T> = ResultValue(value = value, ok = true)
        fun <T> err(message: String): ResultValue<T> = ResultValue(error = message, ok = false)
    }
}

object ResultUtils {
    fun <T> isOk(result: ResultValue<T>): Boolean = result.ok
    fun <T> isErr(result: ResultValue<T>): Boolean = !result.ok
    fun <T> unwrapOr(result: ResultValue<T>, defaultValue: T): T = if (result.ok) result.value!! else defaultValue
    fun <T, U> map(result: ResultValue<T>, mapper: (T) -> U): ResultValue<U> {
        return if (result.ok) ResultValue.ok(mapper(result.value!!)) else ResultValue.err(result.error ?: "error")
    }
}
